tls/extensions/s2n_extension_list.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_extension_list.h"
+#include "s2n_extension_type.h"
+#include "s2n_extension_type_lists.h"
+
+#include <s2n.h>
+
+#include "error/s2n_errno.h"
+#include "utils/s2n_safety.h"
+
+#define s2n_parsed_extension_is_valid(parsed_extension) (parsed_extension->extension.data != NULL)
+
+int s2n_extension_list_send(s2n_extension_list_id list_type, struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+}
+
+int s2n_extension_list_recv(s2n_extension_list_id list_type, struct s2n_connection *conn, struct s2n_stuffer *in)
+{
+    s2n_parsed_extensions_list parsed_extension_list = { 0 };
+    GUARD(s2n_extension_list_parse(in, &parsed_extension_list));
+    GUARD(s2n_extension_list_process(list_type, conn, &parsed_extension_list));
+    return S2N_SUCCESS;
+}
+
+int s2n_extension_process(const s2n_extension_type *extension_type, struct s2n_connection *conn,
+        s2n_parsed_extensions_list *parsed_extension_list)
+{
+    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+}
+
+int s2n_extension_list_process(s2n_extension_list_id list_type, struct s2n_connection *conn,
+        s2n_parsed_extensions_list *parsed_extension_list)
+{
+    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+}
+
+static int s2n_extension_parse(struct s2n_stuffer *in, s2n_parsed_extension *parsed_extensions)
+{
+    notnull_check(parsed_extensions);
+
+    uint16_t extension_type;
+    S2N_ERROR_IF(s2n_stuffer_read_uint16(in, &extension_type) != S2N_SUCCESS,
+            S2N_ERR_BAD_MESSAGE);
+
+    uint16_t extension_size;
+    S2N_ERROR_IF(s2n_stuffer_read_uint16(in, &extension_size) != S2N_SUCCESS,
+            S2N_ERR_BAD_MESSAGE);
+
+    uint8_t *extension_data = s2n_stuffer_raw_read(in, extension_size);
+    S2N_ERROR_IF(extension_data == NULL, S2N_ERR_BAD_MESSAGE);
+
+    s2n_extension_type_id extension_id;
+    if (s2n_extension_supported_iana_value_to_id(extension_type, &extension_id) != S2N_SUCCESS) {
+        /* Ignore unknown extensions */
+        return S2N_SUCCESS;
```

Comment:
I think explicitly setting
```c
*parsed_extensions = NULL
```
will make sure the caller has a `NULL`. Otherwise i may be an uninitialized or reused value, which would probably not be caught.

Reply:
I should probably just clear the whole array at the beginning of the method.

I can't clear the specific parsed extension here, because the specific parsed extension doesn't exist :) There are 2^16 possible extensions, so the parsed extension array only allocates space for the dozen or so that we support.

Reply:
Ah, I misread what was happening. Ignore this :)